### PR TITLE
kvserver: emit MVCC range tombstones over rangefeeds

### LINKED ISSUE #82449: kvserver: emit MVCC range tombstones across rangefeeds
Once MVCC range tombstones are implemented in #70412, and used in `DeleteRange` in #70415, they should be emitted across rangefeeds. These will be recorded as a new `MVCCDeleteRangeOpType` logical op type, and emitted as a new `RangeFeedDeleteRange` rangefeed event. These will be emitted whenever `DeleteRange` is called with the `UseExperimentalRangeTombstone` parameter, and also during catchup scans, but not during initial scans (which never emit tombstones).

Epic CRDB-2624

Jira issue: CRDB-16401

### PR DESCRIPTION
**Pull request by erikgrinaker**
This patch adds MVCC range tombstone support in rangefeeds. Whenever an
MVCC range tombstone is written, a new `MVCCDeleteRangeOp` logical op
is recorded and emitted across the rangefeed as a `RangeFeedDeleteRange`
event. MVCC range tombstones will only be written when the
`MVCCRangeTombstones` version gate has been enabled.

Changefeeds will emit an error for these events. We do not expect to see
these in online spans with changefeeds, since they are initially only
planned for use with schema GC and import rollbacks.

The rangefeed client library has been extended with support for these
events, but no existing callers handle them for the same reason as
changefeeds. Initial scans do not emit regular tombstones, and thus not
range tombstones either, but catchup scans will emit them if
encountered.

This patch has rudimentary testing of MVCC range tombstones in
rangefeeds. A later patch will add a data-driven test harness for
rangefeeds with more exhaustive tests.

Resolves #82449.
Touches #70433.

Release note: None

### CODE REVIEW

**Comment by erikgrinaker** [comment]
This has a pretty hefty performance regression even in the no-range-keys case, mostly due to known performance issues with Pebble for this case. There is also some known overhead with `HasPointAndRange()`. I'm inclined to wait until the optimization work for #82045 lands, which will likely address most of this, and then address the rest of the (then hopefully minor) regression post-merge -- that PR is also required for WithDiff correctness.

```
name                                                 old time/op  new time/op  delta
CatchUpScan/mixed-case/withDiff=true/perc=0.00-24     517ms ± 1%   543ms ± 1%  +4.98%  (p=0.000 n=10+10)
CatchUpScan/mixed-case/withDiff=true/perc=50.00-24    1.03s ± 0%   1.13s ± 0%  +9.28%  (p=0.000 n=10+9)
CatchUpScan/mixed-case/withDiff=true/perc=75.00-24    901ms ± 1%   984ms ± 0%  +9.20%  (p=0.000 n=10+9)
CatchUpScan/mixed-case/withDiff=true/perc=95.00-24    186ms ± 1%   202ms ± 1%  +8.79%  (p=0.000 n=10+10)
CatchUpScan/mixed-case/withDiff=true/perc=99.00-24    157ms ± 1%   170ms ± 1%  +8.29%  (p=0.000 n=10+9)
CatchUpScan/mixed-case/withDiff=false/perc=0.00-24    771ms ± 1%   825ms ± 1%  +6.98%  (p=0.000 n=10+10)
CatchUpScan/mixed-case/withDiff=false/perc=50.00-24   659ms ± 1%   706ms ± 1%  +7.07%  (p=0.000 n=10+10)
CatchUpScan/mixed-case/withDiff=false/perc=75.00-24   590ms ± 1%   632ms ± 1%  +7.17%  (p=0.000 n=10+10)
CatchUpScan/mixed-case/withDiff=false/perc=95.00-24   170ms ± 1%   184ms ± 0%  +8.29%  (p=0.000 n=9+10)
CatchUpScan/mixed-case/withDiff=false/perc=99.00-24   153ms ± 1%   167ms ± 1%  +9.01%  (p=0.000 n=10+10)
```

**Comment by erikgrinaker** [comment]
@miretskiy @stevendanna Feel free to have an early look. This is blocked on a few other PRs, but should otherwise be ~ready.

File: pkg/ccl/streamingccl/streamproducer/event_stream.go
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamproducer/event_stream.go:475]: We would want to forward those, right? Do we need to trim it? Perhaps file an issue to describe how these these should be handled.  Pr

File: pkg/kv/kvserver/rangefeed/catchup_scan.go
Comment (miretskiy) [inline pkg/kv/kvserver/rangefeed/catchup_scan.go:104]: It's strange seeing this since above we say events emitted in chronological order...

File: pkg/kv/kvserver/rangefeed/catchup_scan.go
Comment (miretskiy) [inline pkg/kv/kvserver/rangefeed/catchup_scan.go:155]: is there a possibility of this array being huge (memory blowup?)

Comment (erikgrinaker) [reply pkg/kv/kvserver/rangefeed/catchup_scan.go:155]: There's an implicit assumption that the number of overlapping range keys will be small. Pebble pulls them all into memory at a given key position, as does the higher layers.

File: pkg/ccl/streamingccl/streamproducer/event_stream.go
Comment (erikgrinaker) [reply pkg/ccl/streamingccl/streamproducer/event_stream.go:475]: I don't believe we need to handle them in CDC currently, because we'll only drop these across offline spans. They're trimmed to the registration bounds. For C2C, they'll need to be ingested into the target cluster.

If we're going to use these in online spans, we'd probably want to emit them as distinct events over CDC. Alternatively, we could synthesize point deletion events for any deleted points, but a single range deletion can expand into an unbounded number of point events, which seems bad.

File: pkg/kv/kvserver/rangefeed/catchup_scan.go
Comment (erikgrinaker) [reply pkg/kv/kvserver/rangefeed/catchup_scan.go:104]: Keeping strict chronological ordering below range tombstones is going to kill performance. Consider the following example:

```
5   a5  b5  c5
4   o-----------o
3   a3  b3  c3
2   o-----------o
1   a1  b1  c1
    a   b   c   d
```

If we were to keep strict ordering, then we would have to emit `a1,b1,c1` first, then `[a-d)@2`, then `a3,b3,c3`, and so on. Going back and forth to scan these slices is essentially going to be random access, which is extremely inefficient. Instead, range keys are emitted before the point keys they cover, which is the optimally efficient way to process them -- e.g. `[a-d)@2,[a-d)@4,a1,a3,a5,b1,b3,b5,...`.

**Comment by erikgrinaker** [comment]
This should be ready for final review. Sorry about having to pull in #82045, hopefully won't be too noisy.

**Review by aliher1911 — APPROVED** [review]
Some minor nits, otherwise :lgtm:

File: pkg/storage/mvcc_key.go
Comment (erikgrinaker) [reply pkg/storage/mvcc_key.go:453]: Yeah, good point. I was thinking that given the literal meaning of `HasRangeKeyBetween()`, it doesn't really matter which order the timestamps are in, but it very much does for e.g. range tombstones. I've changed it now so that it simply returns `false` if they're in the wrong order, and added a panic assertion when running under race.

**Comment by erikgrinaker** [comment]
CI failure appears unrelated. TFTR!

bors r=aliher1911

**Pull request merged**
